wp-includes/mysql/class-wp-mysql-lexer.php

```diff
@@ -0,0 +1,2898 @@
+				// Exclude surrogate range U+D800 to U+DFFF:
+				&& ! ( 0xED === $byte_1 && $byte_2 >= 0xA0 )
+				// Exclude overlong encodings:
+				&& ! ( 0xE0 === $byte_1 && $byte_2 < 0xA0 )
+			) {
+				$this->bytes_already_read += 3;
+				continue;
+			}
+
+			// Not a valid identifier character.
+			break;
+		}
+
+		// An identifier cannot consist solely of digits, but we don't need to
+		// ensure that explicitly, as numbers are processed before identifiers.
+
+		return $this->bytes_already_read - $started_at > 0
+			? self::IDENTIFIER
+			: self::INVALID_INPUT;
+	}
+
+	private function read_number(): int {
+		// @TODO: Support numeric-only identifier partes after "." (e.g., 1ea10.1).
+
+		$byte      = $this->sql[ $this->bytes_already_read ] ?? null;
+		$next_byte = $this->sql[ $this->bytes_already_read + 1 ] ?? null;
+
+		if (
+			// HEX number in the form of 0xN.
+			( '0' === $byte && 'x' === $next_byte )
+			// HEX number in the form of x'N' or X'N'.
+			|| ( ( 'x' === $byte || 'X' === $byte ) && "'" === $next_byte )
+		) {
+			$is_quoted                 = "'" === $next_byte;
+			$this->bytes_already_read += 2; // Consume "0x" or "x'".
+			$this->bytes_already_read += strspn( $this->sql, '0123456789abcdefABCDEF', $this->bytes_already_read );
+			if ( $is_quoted ) {
+				$this->bytes_already_read += 1; // Consume the "'".
+			}
+			$type = self::HEX_NUMBER;
+		} elseif (
+			// BIN number in the form of 0bN.
+			( '0' === $byte && 'b' === $next_byte )
+			// BIN number in the form of b'N' or B'N'.
+			|| ( ( 'b' === $byte || 'B' === $byte ) && "'" === $next_byte )
+		) {
+			$is_quoted                 = "'" === $next_byte;
+			$this->bytes_already_read += 2; // Consume "0b" or "b'".
+			$this->bytes_already_read += strspn( $this->sql, '01', $this->bytes_already_read );
+			if ( $is_quoted ) {
+				$this->bytes_already_read += 1; // Consume the "'".
+			}
+			$type = self::BIN_NUMBER;
+		} else {
+			// Here, we have a sequence starting with N or .N, where N is a digit.
+
+			// 1. Try integer first.
+			$this->bytes_already_read += strspn( $this->sql, self::DIGIT_MASK, $this->bytes_already_read );
+			$type                      = self::INT_NUMBER;
+
+			// 2. In case of N. or .N, it's a decimal or float number.
+			if ( '.' === ( $this->sql[ $this->bytes_already_read ] ?? null ) ) {
+				$this->bytes_already_read += 1;
+				$type                      = self::DECIMAL_NUMBER;
+				$this->bytes_already_read += strspn( $this->sql, self::DIGIT_MASK, $this->bytes_already_read );
+			}
+
+			// 3. When exponent is present, it's a float number.
+			$byte         = $this->sql[ $this->bytes_already_read ] ?? null;
+			$next_byte    = $this->sql[ $this->bytes_already_read + 1 ] ?? null;
+			$has_exponent =
+				( 'e' === $byte || 'E' === $byte )
+				&& (
+					$this->is_digit( $next_byte )
+					|| (
+						( '+' === $next_byte || '-' === $next_byte )
+						&& $this->is_digit( $this->sql[ $this->bytes_already_read + 2 ] ?? null )
```

Comment:
Or check `strlen()` explicitly instead of doing `?? null`. The reader needs to translate the latter into "ooh, so we pass null when the string is too short" anyway – let's do that work for them in the code. As a bonus, this will enable inlining the is_digit() method as we won't have to worry about receiving a `null` argument.

Reply:
Ditto for any method where we do `return null !== $byte`